### ISSUE TITLE
`vagrant global-status --prune` can't find Homestead settings file

### ISSUE THREAD

**Issue by sebastiaanluca** [laravel/homestead#716]
It seems either PR https://github.com/laravel/homestead/pull/252 doesn't work anymore or I'm doing something wrong 😊

## Expected behavior

Vagrant deletes all out-of-date boxes and virtual machines.

## Actual behavior

- An error is thrown when running `vagrant box prune` __inside__ a project that has Homestead setup or a Vagrant box
- `vagrant global-status --prune` fails to complete (when running anywhere)

## Steps to reproduce

Execute `vagrant global-status --prune` or `vagrant box prune`.

## Versions

- Vagrant: 1.9.7
- Provider: 5.1.22
- Homestead: 6.5.0

I have it both on an old and a new project. I've upgraded the old project to Homestead 6.5, but it still won't pass.

## Host operating system

macOS 10.12.6

## Vagrantfile

Basic Homestead setup with an unchanged Vagrantfile except for the paths.

```
# -*- mode: ruby -*-
# vi: set ft=ruby :

require 'json'
require 'yaml'

VAGRANTFILE_API_VERSION ||= "2"
confDir = $confDir ||= File.expand_path("vendor/laravel/homestead", File.dirname(__FILE__))

homesteadYamlPath = File.expand_path("Homestead.yaml", File.dirname(__FILE__))
homesteadJsonPath = File.expand_path("Homestead.json", File.dirname(__FILE__))
afterScriptPath = "scripts/vm/provision.sh"
aliasesPath = "scripts/vm/aliases.sh"

require File.expand_path(confDir + '/scripts/homestead.rb')

Vagrant.require_version '>= 1.9.0'

Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
    if File.exist? aliasesPath then
        config.vm.provision "file", source: aliasesPath, destination: "/tmp/bash_aliases"
        config.vm.provision "shell" do |s|
            s.inline = "awk '{ sub(\"\r$\", \"\"); print }' /tmp/bash_aliases > /home/vagrant/.bash_aliases"
        end
    end

    if File.exist? homesteadYamlPath then
        settings = YAML::load(File.read(homesteadYamlPath))
    elsif File.exist? homesteadJsonPath then
        settings = JSON.parse(File.read(homesteadJsonPath))
    else
        abort "Homestead settings file not found in #{confDir}"
    end

    Homestead.configure(config, settings)

    if File.exist? afterScriptPath then
        config.vm.provision "shell", path: afterScriptPath, privileged: false
    end

    if defined? VagrantPlugins::HostsUpdater
        config.hostsupdater.aliases = settings['sites'].map { |site| site['map'] }
    end
end
```

## References

- https://github.com/laravel/homestead/pull/252
- https://github.com/laravel/homestead/pull/252/commits/017232ccb5d3ebb5ec9a19210f93c0189189f70a

**Comment by svpernova09** [comment]
What's the error your getting?

What's the output of `vagrant global-status`

What does `unchanged Vagrantfile except for the paths.` mean? You shouldn't have to edit `Vagrantfile` at all.

**Comment by sebastiaanluca** [comment]
### Error

It's identical to https://github.com/laravel/homestead/pull/252#issue-102096039, so what happens:

`vagrant global-status --prune` (executed from anywhere):

```
Homestead settings file not found in /Users/Sebastiaan/[directoryofrandomprojectusinghomestead]/vendor/laravel/homestead
```

`vagrant box prune` inside a folder with Vagrant set up:

```
The following boxes will be kept...
geerlingguy/ubuntu1604 (virtualbox, 1.1.6)
laravel/homestead      (virtualbox, 4.0.0)
ubuntu/trusty64        (virtualbox, 20171030.0.1)
ubuntu/xenial64        (virtualbox, 20171028.0.0)

Checking for older boxes...
Homestead settings file not found in /Users/Sebastiaan/[currentprojectdirectory]/vendor/laravel/homestead
```

`vagrant box prune` inside a folder with no Vagrantfile or .vagrant directory: deletes all old boxes as expected (no errors).

### Global status

`vagrant global-status` shows me all the virtual machines without errors.

```
id       name                provider   state    directory
--------------------------------------------------------------------------------------------------------------------------------
4a84e20  [redacted] virtualbox poweroff /Users/Sebastiaan/[redacted]
c9a5b98  [redacted]          virtualbox poweroff /Users/Sebastiaan/Workspace/[redacted]
73995ca  homestead           virtualbox poweroff /Users/Sebastiaan/Homestead
34497ff  [redacted]                virtualbox poweroff /Users/Sebastiaan/[redacted]
22a2492  laravel-skeleton    virtualbox poweroff /Users/Sebastiaan/[redacted]
601d595  default             virtualbox poweroff /Users/Sebastiaan/[redacted]
ff27d5a  [redacted]                 virtualbox running  /Users/Sebastiaan/[redacted]
```

### Vagrantfile

> What does unchanged Vagrantfile except for the paths. mean?

I always change the before and after script paths so it uses my own scripts to provision the virtual machine. Nothing else in the file is changed and it uses the referenced fix, so it shouldn't throw an error when trying to prune/list the box/VM.

**Comment by svpernova09** [comment]
I'm not sure why this is happening other than vagrant just being odd about relative paths in our configs (which we use ruby to combine into full paths).

Aside from the error message (which seems rather innocuous to me) what are you being prevented from doing? Cleaning up all the vagrant boxes laying around the filesystem? You can always just open up Virtual box and delete the VMs there (make sure to select `delete all files`) to accomplish similar.

**Comment by sebastiaanluca** [comment]
Yeah, it somehow stumbles on the relativity of the paths. Vagrantfile seems to look for the Homestead config in its own composer vendor directory instead of 3 levels up (project root). Though the PR fixed that, but must be something else still in there that uses relative paths.

Manually deleting machines would be a hassle though. Both prune commands are very easy in that they check which machine or box is still in use and warns me about that before deleting it —which is why I want to use those commands. This bug prevents me from cleaning up boxes/machines and thus valuable disk space :p There's also a Vagrant "bug" where a 'default' machine is listed, but doesn't exist anymore. Only pruning all machines seems to solve that.

**Comment by svpernova09** [comment]
Yeah I totally get it. Closing this for now, will investigate but unsure of what a Homestead code change to fix this may be.

**Comment by sebastiaanluca** [comment]
Hey guys, any update on this? Can't remove any boxes through vagrant now either. Same error.

**Comment by svpernova09** [comment]
Open virtual box stop any vagrant boxes. Delete any you want to, remove all files.

--
- Joe Ferguson
JoeFerguson.me
MemphisPHP.org

**Comment by sebastiaanluca** [comment]
That's a somewhat fine workaround for deleting active virtual machines, but I want to delete outdated vagrant boxes. The only way to know which ones are outdated ánd aren't used by any virtual machine is by running the prune and/or outdated command, but that doesn't work because it halts on `Homestead settings file not found in …`. So basically those (and perhaps other) vagrant commands are broken if you have at least one Homestead VM.

I'd be happy to debug and maybe even fix this myself, but I have no idea where to look or what the issue is.

**Comment by svpernova09** [comment]
Show me the full output of `vagrant global-status`

**Comment by sebastiaanluca** [comment]
Just a list of all my current and previous virtual machines, including those that don't exist anymore.

```
$ vagrant global-status

id       name                  provider   state    directory
----------------------------------------------------------------------------------------------------------------------------------
f3e064d  default               virtualbox poweroff /Users/Sebastiaan/Workspace/Projects/redacted
c9a5b98  redacted              virtualbox poweroff /Users/Sebastiaan/Workspace/Projects/redacted
73995ca  homestead             virtualbox poweroff /Users/Sebastiaan/Homestead
22a2492  redacted              virtualbox poweroff /Users/Sebastiaan/Workspace/Projects/redacted
601d595  default               virtualbox poweroff /Users/Sebastiaan/Workspace/Projects/redacted
b1070ae  homestead-7           virtualbox poweroff /Users/Sebastiaan/Workspace/Projects/redacted
9ae2371  default               virtualbox poweroff /Users/Sebastiaan/Workspace/Projects/redacted
3d5c50b  redacted              virtualbox poweroff /Users/Sebastiaan/Workspace/Projects/redacted
6becafe  redacted              virtualbox running  /Users/Sebastiaan/Workspace/Projects/redacted
8ad6c1a  redacted              virtualbox running  /Users/Sebastiaan/Workspace/Projects/redacted

The above shows information about all known Vagrant environments
on this machine. This data is cached and may not be completely
up-to-date (use "vagrant global-status --prune" to prune invalid
entries). To interact with any of the machines, you can go to that
directory and run Vagrant, or you can use the ID directly with
Vagrant commands from any directory. For example:
"vagrant destroy 1a2b3c4d"
```

```
$ vagrant box list

debian/stretch64       (virtualbox, 9.4.0)
geerlingguy/ubuntu1604 (virtualbox, 1.1.6)
laravel/homestead      (virtualbox, 2.1.0)
laravel/homestead      (virtualbox, 5.2.0)
laravel/homestead      (virtualbox, 6.1.0)
ubuntu/trusty64        (virtualbox, 20171030.0.1)
ubuntu/xenial64        (virtualbox, 20171028.0.0)
```

```
$ vagrant box prune

The following boxes will be kept...
debian/stretch64       (virtualbox, 9.4.0)
geerlingguy/ubuntu1604 (virtualbox, 1.1.6)
laravel/homestead      (virtualbox, 6.1.0)
ubuntu/trusty64        (virtualbox, 20171030.0.1)
ubuntu/xenial64        (virtualbox, 20171028.0.0)

Checking for older boxes...
Homestead settings file not found in /Users/Sebastiaan/Workspace/Projects/sebastiaanluca.com-vm/vendor/laravel/homestead
```

```
$ vagrant box remove laravel/homestead --box-version=2.1.0

Homestead settings file not found in /Users/Sebastiaan/Workspace/Projects/redacted/vendor/laravel/homestead
```

**Comment by svpernova09** [comment]
> 
```
$ vagrant box remove laravel/homestead --box-version=2.1.0

Homestead settings file not found in /Users/Sebastiaan/Workspace/Projects/redacted/vendor/laravel/homestead
```

Run this:
```
cd /Users/Sebastiaan/Workspace/Projects/redacted
rm -rf .vagrant

**Comment by svpernova09** [comment]
If the issue is still there after you do ^ 

The next step is to find all vagrant machines everywhere on your system show the entire output of `sudo find / -iname .vagrant`

**Comment by sebastiaanluca** [comment]
Uhm, every time I run that remove command it's a different project. Usually the last VM that got booted. Not going to delete active VM's 😅

So IMO it just reads the Vagrantfile of every vagrant VM and fails because of those relative paths or something similar. So not sure what your goal is trying to find all machines on my system?

**Comment by svpernova09** [comment]
> Uhm, every time I run that remove command it's a different project. Usually the last VM that got booted. Not going to delete active VM's 😅

https://github.com/laravel/homestead/issues/909#issuecomment-407208620

```
* Let this be the last time you put something IN a VM you don't have a backup of
```

> So not sure what your goal is trying to find all machines on my system?

The idea would be to go delete ALL those `.vagrant` folders of any paths NOT found in the output of `vagrant global-status`

**Comment by sebastiaanluca** [comment]
I only have test data in my VMs ;) Just had an older Homestead installation that I don't have the test data for anymore, but I managed to make a backup of the existing data.

I ended up purging everything related to Vagrant and Homestead from my system to start anew. No boxes, no VMs, no .vagrant directories left anywhere. Since I only have one box and 2 identical VMs now, the commands seem to work again without errors. Will test again when I have some more from other projects.

Thanks for the help!
